cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,680 @@
+               "columns' metadata should be equal to the number of columns in table");
+
+  nanoarrow::UniqueSchema result;
+  ArrowSchemaInit(result.get());
+  NANOARROW_THROW_NOT_OK(ArrowSchemaSetTypeStruct(result.get(), input.num_columns()));
+
+  for (int i = 0; i < input.num_columns(); ++i) {
+    auto child = result->children[i];
+    auto col   = input.column(i);
+    ArrowSchemaInit(child);
+    NANOARROW_THROW_NOT_OK(ArrowSchemaSetName(child, metadata[i].name.c_str()));
+    child->flags = col.has_nulls() ? ARROW_FLAG_NULLABLE : 0;
+
+    if (col.type().id() == cudf::type_id::EMPTY) {
+      NANOARROW_THROW_NOT_OK(ArrowSchemaSetType(child, NANOARROW_TYPE_NA));
+      continue;
+    }
+
+    NANOARROW_THROW_NOT_OK(
+      cudf::type_dispatcher(col.type(), detail::dispatch_to_arrow_type{}, col, metadata[i], child));
+  }
+
+  auto out = std::make_unique<ArrowSchema>();
+  result.move(out.get());
+  return out;
+}
+
+std::unique_ptr<ArrowDeviceArray> to_arrow_device(cudf::table&& table,
+                                                  rmm::cuda_stream_view stream,
+                                                  rmm::mr::device_memory_resource* mr)
+{
+  nanoarrow::UniqueArray tmp;
+  NANOARROW_THROW_NOT_OK(ArrowArrayInitFromType(tmp.get(), NANOARROW_TYPE_STRUCT));
+
+  NANOARROW_THROW_NOT_OK(ArrowArrayAllocateChildren(tmp.get(), table.num_columns()));
+  tmp->length     = table.num_rows();
+  tmp->null_count = 0;
+
+  auto cols = table.release();
+  for (size_t i = 0; i < cols.size(); ++i) {
+    auto child = tmp->children[i];
+    auto col   = cols[i].get();
+
+    if (col->type().id() == cudf::type_id::EMPTY) {
+      NANOARROW_THROW_NOT_OK(ArrowArrayInitFromType(child, NANOARROW_TYPE_NA));
+      child->length     = col->size();
+      child->null_count = col->size();
+      continue;
+    }
+
+    NANOARROW_THROW_NOT_OK(cudf::type_dispatcher(
+      col->type(), detail::dispatch_to_arrow_device{}, std::move(*col), stream, mr, child));
+  }
+
+  NANOARROW_THROW_NOT_OK(
+    ArrowArrayFinishBuilding(tmp.get(), NANOARROW_VALIDATION_LEVEL_MINIMAL, nullptr));
+
+  auto private_data = std::make_unique<detail::ArrowDeviceArrayPrivateData>();
+  cudaEventCreate(&private_data->sync_event);
+
+  auto status = cudaEventRecord(private_data->sync_event, stream);
+  if (status != cudaSuccess) { CUDF_FAIL("could not create event to sync on"); }
+
+  ArrowArrayMove(tmp.get(), &private_data->parent);
+  auto result       = std::make_unique<ArrowDeviceArray>();
+  result->device_id = rmm::get_current_cuda_device().value();
+  // can/should we check whether the memory is managed/cuda_host memory?
+  result->device_type        = ARROW_DEVICE_CUDA;
+  result->sync_event         = &private_data->sync_event;
+  result->array              = private_data->parent;
+  result->array.private_data = private_data.release();
+  result->array.release      = &detail::ArrowDeviceArrayRelease;
+  return result;
```

Comment:
@paleolimbot got me thinking about exactly what this API should look like in [this thread](https://github.com/apache/arrow-nanoarrow/pull/406#discussion_r1533130121). Right now, we're returning a `std::unique_ptr<ArrowDeviceArray>`, but we're effectively assuming that the consumer will take this and follow the C Data Interface rules. If the consumer does nothing at all with the output, then this would constitute a leak because the release hook would never have been called. Therefore, the way I see it we have two options that would be more internally consistent than what we do now:
- Construct this `unique_ptr` with a suitable deleter so that the `unique_ptr` handles the release properly even if the consumer does not. Then, the consumer is responsible for calling `release` on the returned `unique_ptr` to signal taking ownership of the array.
- Return a raw pointer to signal to the consumer that all memory management of this object is now their concern. This would be unsafe, but would at least be a little more honest than the current approach. IMO this is definitely the inferior of the two options, but I wanted to put it out there.

Does that assessment seem fair to you? Also definitely curious to hear Dewey's take here.

Reply:
@vyasr I chose to have it return the unique_ptr based on our conversations, but like @paleolimbot said, the common paradigm on the Arrow side is to accept a pointer output argument. Do you have a preference between setting the proper deleter vs taking an output pointer as an argument?

Reply:
I'm inclined to keep the unique_ptr and set the deleter. I think this is cleaner solution and more consistent with most libcudf APIs in general. I'd rather not start a pattern where pointers are provided for the API to fill in if possible. So given the choice I prefer a more C++ STL solution than a C one.

Reply:
Sure thing. I'll update to use a custom deleter with the unique ptrs then